esda/shape.py

```diff
@@ -0,0 +1,332 @@
+# -------------------- INTERTIAL MEASURES -------------------- #
+
+
+def moment_of_inertia(collection):
+    """
+    Computes the moment of inertia of the chain.
+
+    This treats each boundary point as a point-mass of 1.
+
+    Thus, for constant unit mass at each boundary point,
+    the MoI of this pointcloud is
+
+    \sum_i d_{i,c}^2
+
+    where c is the centroid of the polygon
+
+    Altman's OS_1 measure, cited in Boyce and Clark (1964), also used in Weaver
+    and Hess (1963).
+    """
+    coords = pygeos.get_coordinates(collection)
+    geom_ixs = numpy.tile(
+        numpy.arange(len(collection)), pygeos.get_num_coordinates(collection)
+    )
+    centroids = pygeos.get_coordinates(pygeos.centroid(collection))[geom_ixs]
+    squared_euclidean = numpy.sum((coords - centroids) ** 2, axis=1)
+    dists = (
+        pandas.DataFrame.from_dict(dict(d2=squared_euclidean, geom_ix=geom_ixs))
+        .groupby("geom_ix")
+        .d2.sum()
+    ).values
+    return pygeos.area(collection) / numpy.sqrt(2 * dists)
+
+
+def moa_ratio(collection):
+    """
+    Computes the ratio of the second moment of area (like Li et al (2013)) to
+    the moment of area of a circle with the same area.
+    """
+    raise NotImplementedError()
+
+
+def nmi(collection):
+    """
+    Computes the Normalized Moment of Inertia from Li et al (2013), recognizing
+    that it is the relationship between the area of a shape squared divided by
+    its second moment of area.
+    """
+    raise NotImplementedError()
+
+
+def second_moment_of_area(collection):
+    """
+    Using equation listed on en.wikipedia.org/Second_Moment_of_area, the second
+    moment of area is actually the cross-moment of area between the X and Y
+    dimensions:
+
+    I_xy = (1/24)\sum^{i=N}^{i=1} (x_iy_{i+1} + 2*x_iy_i + 2*x_iy_{i+1} +
+    x_{i+1}y_i)(x_iy_i - x_{i+1}y_i)
+
+    where x_i, y_i is the current point and x_{i+1}, y_{i+1} is the next point,
+    and where x_{n+1} = x_1, y_{n+1} = 1.
+
+    This relation is known as the:
+    - second moment of area
+    - moment of inertia of plane area
+    - area moment of inertia
+    - second area moment
+
+    and is *not* the mass moment of inertia, a property of the distribution of
+    mass around a shape.
+    """
+    raise NotImplementedError()
+
+
+# -------------------- OTHER MEASURES -------------------- #
+
+
+def reflexive_angle_ratio(collection):
+    """
+    The Taylor reflexive angle index, measure OS_4 in Altman (1998)
+
+    (N-R)/(N+R), the difference in number between non-reflexive angles and
+    reflexive angles in a polygon, divided by the number of angles in the
+    polygon in general.
+    """
+    angles, geom_indices = get_angles(collection, return_indices=True)
+    return (
+        pandas.DataFrame.from_dict(dict(is_reflex=angles < 0, geom_ix=geom_indices))
+        .groupby("geom_ix")
+        .is_reflex.mean()
+    )
+
+
+def fractal_dimension(collection):
+    """"""
+    raise NotImplementedError()
```

Comment:
This one? https://github.com/martinfleis/momepy/blob/ec7f172806d8390f38ee61104ce1b2a2bab7fd82/momepy/shape.py#L102

Reply:
Yep!

Reply:
Well, I should say, yes I mean the same concept, but no, I think a different estimator is appropriate for data not on grid. 

Going back to the original [Olsen et al. 1993](https://www.asprs.org/wp-content/uploads/pers/1993journal/oct/1993_oct_1517-1520.pdf) I believe that formula only applies to growth on grids, I believe (hence the .25). Assuming another gridding system gives different results. 

[Another paper has done this](https://doi.org/10.1155/2020/7528703) but I'm sketched out by that publisher, so I want to validate it.

Reply:
Seems like
1.  this should get re-classed into the "ratio" measures unless we implement some kind of box-counting or varying-ruler method. Using the Olsen et al. (1993) derivation, you start from an ideal shape, then derive scaling constants based on that relation... that makes it an ideal shape measure. 
2. using a grid versus a hex versus a circle makes very little difference in the shapes I'm seeing, although it consistently underestimates the values. I'll add all three and make hex the default.

Reply:
The one in momepy comes from [this paper](https://doi.org/10.1016/j.landurbplan.2012.02.008) where it was used on polygons. But looking into the original references you are probably right that it comes from grid world.

Reply:
Wow, for the test shape, you get basically invalid results for these shapes from these relations... the circle and hex agree pretty strongly (about `.2`) but the grid formula is negative. 

Anyway, we may need to caveat this in the docstring about very simple shapes, as it seems this is not the best way to do this when dealing with polygons in vector space.

Reply:
> The one in momepy comes from this paper where it was used on polygons. But looking into the original references you are probably right that it comes from grid world.

hmm yeah. I bet it's fine for square-like things, but for arbitrary polygons, I would definitely be careful with it. 

There are [a lot of ways to estimate fractal dimension](https://doi.org/10.1007/BF02065874):

![Screenshot 2021-04-28 at 10 19 02](https://user-images.githubusercontent.com/2250995/116379886-5b9c2b00-a80b-11eb-8915-0cdba2e71c7a.png)

I bet there are all kinds of pitfalls here, so I'll add something to the docstring. Don't really have the time to look into this right now myself.